predicate.go

```diff
@@ -0,0 +1,219 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api/v1"
+)
+
+// Predicates implemented filter functions.
+// The filter list is expected to be a subset of the supplied list.
+func Predicates(clientset *kubernetes.Clientset) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var buf bytes.Buffer
+		body := io.TeeReader(r.Body, &buf)
+		glog.V(2).Infof("body: %s", buf.String())
+
+		var extenderArgs schedulerapi.ExtenderArgs
+		var extenderFilterResult *schedulerapi.ExtenderFilterResult
+
+		if err := json.NewDecoder(body).Decode(&extenderArgs); err != nil {
+			extenderFilterResult = &schedulerapi.ExtenderFilterResult{
+				Nodes:       nil,
+				FailedNodes: nil,
+				Error:       err.Error(),
+			}
+		} else {
+			extenedResourceScheduler := &ExtendedResourceScheduler{
+				Clientset: clientset,
+			}
+			extenderFilterResult = filter(extenderArgs, extenedResourceScheduler)
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		if resultBody, err := json.Marshal(extenderFilterResult); err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Write([]byte(err.Error()))
+		} else {
+			w.WriteHeader(http.StatusOK)
+			w.Write(resultBody)
+		}
+	}
+}
+
+// Bind delegates the action of binding a pod to a node to the extender.
+func Bind(clientset *kubernetes.Clientset) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var buf bytes.Buffer
+		body := io.TeeReader(r.Body, &buf)
+		var extenderBindingArgs schedulerapi.ExtenderBindingArgs
+
+		if err := json.NewDecoder(body).Decode(&extenderBindingArgs); err != nil {
+			glog.Errorf("json decode error: %v", err)
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(err.Error()))
+		}
+		b := &v1.Binding{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: extenderBindingArgs.PodNamespace,
+				Name:      extenderBindingArgs.PodName,
+				UID:       extenderBindingArgs.PodUID,
+			},
+			Target: v1.ObjectReference{
+				Kind: "Node",
+				Name: extenderBindingArgs.Node,
+			},
+		}
+		err := clientset.CoreV1().Pods(b.Namespace).Bind(b)
+		if err != nil {
+			glog.Errorf("bind failed: %v", err)
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(err.Error()))
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func filter(extenderArgs schedulerapi.ExtenderArgs, extenedResourceScheduler *ExtendedResourceScheduler) *schedulerapi.ExtenderFilterResult {
```

Comment:
do you really know how to filter ERs for ERC ?

Reply:
I don't understand here